### ISSUE TITLE
High CPU usage in browser

### ISSUE THREAD

**Issue by willemneal** [orbitdb/orbitdb#353]
I've been extending the browser example to create a notepad app.  I've found though that tabs running the app use a lot of CPU up to 200%.

From using the javascript profiler, 68% of the time is spent here:

```javascript
exports.createUnboxStream = (decipher, mac) => {
  return pull(
    ensureBuffer(),
    lp.decode(lpOpts),
    pull.asyncMap((chunk, cb) => {
      const l = chunk.length
      const macSize = mac.length

      if (l < macSize) {
        return cb(new Error(`buffer (${l}) shorter than MAC size (${macSize})`))
      }

      const mark = l - macSize
      const data = chunk.slice(0, mark)
      const macd = chunk.slice(mark)

      mac.digest(data, (err, expected) => {
        if (err) {
          return cb(err)
        }

        if (!macd.equals(expected)) {
          return cb(new Error(`MAC Invalid: ${macd.toString('hex')} != ${expected.toString('hex')}`))
        }

        // all good, decrypt
        decipher.decrypt(data, (err, decrypted) => {
          if (err) {
            return cb(err)
          }

          cb(null, decrypted)
        })
      })
    })
  )
}
```

Any thoughts?

**Comment by fazo96** [comment]
It looks like this code: https://github.com/libp2p/js-libp2p-secio/blob/4d4163def531a7ef6626690f7915521c6b04592e/src/etm.js#L33

It's part of libp2p (used by IPFS which in turn is used by orbit-db) and I think it is used to decrypt incoming messages from other peers.

I guess your libp2p is being super chatty for some reason, are you writing huge amounts of data to the db? Could also be just a bug somewhere in the stack, or maybe it's just very slow, but that high cpu usage looks suspicious

**Comment by willemneal** [comment]
Thanks Enrico,

It happens when the client is in a steady state and not writing anything.
It sounds like, as an IPFS node, the client is receiving a lot of
requests.  Could this be solved by bootstraping to less peers initialy?

**Comment by haadcode** [comment]
Thanks @willemneal for reporting the issue!

High CPU usage doesn't sound normal and in the past that has happened due to various reasons. 

One obvious one we had in the past was that WebRTC was super heavy on resource usage and caused high CPU, so one thing to check is to make sure WebRTC is not used by IPFS (but Websocket), see more info here: https://github.com/ipfs/js-ipfs#is-there-a-more-stable-alternative-to-webrtc-star-that-offers-a-similar-functionality.

You could also try to narrow down the problem by using a [local websocket signal server](https://github.com/libp2p/js-libp2p-websocket-star-rendezvous) which should limit the connections to just the peers you have running/connecting to it.

Another part of puzzle that would be interesting is to see what triggers those calls that are taking the time. If you have time to look into it, sharing a full profiler dump (the file) from Chrome devtools would be great and we can take a look at it to see if there's anything else suspicious going.

Can you also share which OS and browser this is happening and whether this happens on other browsers too?

**Comment by Jon-Biz** [comment]
I'm seeing this same behaviour on 
* osX 10.14.3
* Chrome version 80.0.3987.149 (Official Build) (64-bit)

This is with no nodes connected, fwiw.

**Comment by chrisbarber** [comment]
I'm getting high-ish CPU too (20-40%), also while in idle/steady-state.

>One obvious one we had in the past was that WebRTC was super heavy on resource usage and caused high CPU, so one thing to check is to make sure WebRTC is not used by IPFS (but Websocket)

Would this apply even while "idle" (not writing to any db, but I suppose some communication with the signalling server could still be happening)? I have been going with webrtc currently because of wanting to run more recent ipfs+orbitdb. As far as I can tell websocket was dropped from the newer ipfs for some reason-- referring to here https://github.com/ipfs/js-ipfs/issues/2779#issuecomment-628743366. I guess eventually the star stuff all will be phased out in favor of this rendevous protocol. Just a bit tricky to figure out what we can use in the meantime!

I'm new to trying out all this ipfs / orbit / decentralized eco-system, it is quite fun to try out. Just developing a demo app for which I want to have decentralized syncing between a user's devices.

**Comment by chrisbarber** [comment]
Okay so some findings. I double-checked and tried with a local `libp2p-websocket-star-rendezvous` server, and ipfs comes back with the same error in https://github.com/ipfs/js-ipfs/issues/2779. As far as I can tell this leaves the only working option right now being webrtc-star. I'm not sure whether circuit-relay can be used for generalized discovery; I have not seen it recommended.

I'm on OSX 10.15.4, and chrome 81.0.4044.129 (Official Build) (64-bit). That said I have seen high CPU on my windows+chrome setup as well.

Not doing any db operations besides an initial load of a db which is practically empty, and furthermore I don't have the db open in another instance of the app so there is nowhere it could be replicating with.

Nonetheless there appears to be a huge amount of ws traffic; the following is from just 1 minute of sitting idle. See the megabyte values at the bottom (and... what does it mean? 0 B / 34.1 MB transferred?).

<img width="568" alt="Screenshot 2020-05-30 at 20 57 03" src="https://user-images.githubusercontent.com/1530840/83337917-95914400-a2bf-11ea-939e-7b2b65d4f0c8.png">

The ws connections above are to ipfs bootstrap servers. The traffic continues even if I shut down my local webrtc `star-signal` server.

This profile includes the startup and then 20 seconds or so of steady-state. All the activity appears to be ipfs; I don't think orbitdb is doing anything. I don't know how this profile looks when loading but I unminified orbit and ipfs so maybe it would be more readable. I'm on `ipfs@0.44.0` and `orbit-db@0.24.1`.

[Profile-20200530T214529.json.zip](https://github.com/orbitdb/orbit-db/files/4706321/Profile-20200530T214529.json.zip)

**Comment by chrisbarber** [comment]
Okay with a fresh db this doesn't happen. So I guess oplog had grown long. Still doesn't make sense to me; I was not writing much, why it should take so long to load. I thought the cached oplog would also exist in local storage on the browser so it wouldn't have to be fully downloaded each time or whatever it's trying to do.

**Comment by haadcode** [comment]
There have been various issues with with a) CPU usage in browsers (often related to webrtc) and b) memory usage (related to orbitdb).

These should be fixed in 1.0.0. Note the [new package name](https://www.npmjs.com/package/@orbitdb/core).

If the issues persist, feel free to re-open the issue.
